extensions/cms/src/tasks/export.rs

```diff
@@ -0,0 +1,452 @@
+
+        let primary_purls = match self.service.get_primary_purls().await {
+            Ok(opt) => match opt {
+                Some(purls) => purls,
+                None => {
+                    return Err(Error::Analytic(
+                        "Error attempting to get primary purls, none found!".to_string(),
+                    ))
+                }
+            },
+            Err(err) => {
+                return Err(Error::Analytic(format!(
+                    "Error attempting to get primary purls: {err}"
+                )))
+            }
+        };
+
+        task.count = primary_purls.len() as u64;
+        self.update(task).await?;
+
+        println!("==> processing {} sboms for detail report", task.count);
+        let mut iteration = 0;
+
+        for purl in primary_purls {
+            iteration += 1;
+            println!(
+                "==> generating detail report for iteration {} for purl {}",
+                iteration, purl
+            );
+
+            match self.service.export_by_purl(purl.as_str()).await {
+                Ok(file_path_option) => {
+                    if let Some(file_path) = file_path_option {
+                        println!("==> Sbom detail report complete for {}", purl);
+                        report_paths.push(file_path)
+                    }
+                }
+                Err(err) => {
+                    println!("==> Sbom detail report complete for {}", purl);
+                    errors.insert(purl, format!("{}", err));
+                }
+            }
+        }
+
+        task.status = TaskStatus::Complete;
+
+        Ok(errors)
+    }
+}
+
+impl Service<Task> for ExportTask {
+    fn store(&self) -> Arc<Store> {
+        self.service.store.clone()
+    }
+}
+
+#[derive(Debug)]
+pub struct ExportService {
+    pub(crate) store: Arc<Store>,
+    pub(crate) storage: Arc<dyn StorageProvider>,
+}
+
+impl ExportService {
+    /// Factory method for new instance of type.
+    pub fn new(store: Arc<Store>, storage: Arc<dyn StorageProvider>) -> Self {
+        ExportService { store, storage }
+    }
+
+    /// Queries MongoDB to get all of the purls for the primary SBOMs
+    pub(crate) async fn get_primary_purls(&self) -> Result<Option<Vec<String>>, Error> {
+        let pipeline = Pipeline::new(self.store.clone());
+        pipeline.add_stage(Stage::new(json!({
+            "$match": {
+                "kind": "primary"
+            }
+        })));
+
+        pipeline.add_stage(Stage::new(json!({
+            "$project": {
+              "_id": 0,
+              "purl": 1,
+            }
+        })));
+
+        pipeline.add_stage(Stage::new(json!({
+            "$group": {
+                "_id": "",
+                "purls": {
+                  "$push": {
+                    "purl": "$purl"}
+                    }
+                }
+            }
+        )));
+
+        match pipeline.execute_on("Package").await {
```

Comment:
One improvement I can see here is that we could be using a type rather than a String to specify the collection we want to execute on.  If you are already in this code and it's not too off the grid with respect to the purpose of the ticket, refactoring the api here to be like: 
`pipeline.execute_on::<Package>()` ensures that the collection is, in fact a real collection, makes accidentally misspelling the name of the collection  impossible.  It also makes it easy to rename the collection if necessary because it's only a single refactor of the struct used to rename the Mongo document and we won't accidentally leave the old name of the Document here.

Reply:
I started doing this and quickly determined it was more involved that I'd like given the urgency of the related EPSS issue, so I'm punting on that for now.

Reply:
:+1: